### PR TITLE
Christoph/fix/typehints

### PR DESCRIPTION
**Pull request by cburgdorf**
This is mostly #1404 by @Bhargavasomu with some minimal fine tuning on top.

### What was wrong?
The `eth` module was not `type hinted` till now. This needs to be fixed and also the `tox.ini` file needs to be updated.
Issue : #1398

### How was it fixed?
Adding `type hints` to the arguments and the function. The phase of adding the type hints is as follows
* eth.utils (Started previously and hence wanted to finish it)
* eth.tools
* eth.vm
* Everything else which is leftover

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

File: eth/tools/builder/chain/builders.py [hunk 1]
```diff
@@ -40,11 +77,12 @@
     ConstantinopleVM,
 )
 
-if TYPE_CHECKING:
-    from typing import Dict, Union  # noqa: F401
 
+VMConfigurationType = Iterable[Tuple[int, Type[BaseVM]]]
+GeneralStateType = Union[AccountState, List[Tuple[Address, Dict[str, Union[int, bytes, Dict[int, int]]]]]]  # noqa: E501
```
Comment (pipermerriam) [inline eth/tools/builder/chain/builders.py]: Looks like the inner `Dict` should be a `TypedDict`

Comment (pipermerriam) [inline eth/tools/builder/chain/builders.py]: I think the inner `Dict` should be updated to be an `AccountDetals`.

File: eth/typing.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    Address,
+)
+from mypy_extensions import (
+    TypedDict,
+)
+
+
+# TODO: Move into eth_typing
+
+AccountDetails = TypedDict('AccountDetails',
```
Comment (pipermerriam) [inline eth/typing.py]: Ooooh!  `TypedDict` is cool.

File: eth/utils/module_loading.py [hunk 3]
```diff
@@ -1,8 +1,13 @@
...
 from importlib import import_module
 
+from typing import (
+    Any,
+    Tuple,
+)
 
-def import_string(dotted_path):
+
+def import_string(dotted_path: str) -> Any:
```
Comment (pipermerriam) [inline eth/utils/module_loading.py]: can you use `types.ModuleType` here instead of `Any`?

File: eth/utils/numeric.py [hunk 4]
```diff
@@ -12,7 +15,7 @@
 )
 
 
-def int_to_bytes32(value):
+def int_to_bytes32(value: Union[int, bool]) -> bytes:
```
Comment (pipermerriam) [inline eth/utils/numeric.py]: I think `Hash32` is more appropriate than `bytes`

File: eth/utils/transactions.py [hunk 5]
```diff
@@ -43,7 +47,10 @@ def extract_signature_v(v: int) -> int:
         return V_OFFSET
 
 
-def create_transaction_signature(unsigned_txn, private_key, chain_id=None):
+def create_transaction_signature(unsigned_txn: BaseTransaction,
+                                 private_key: datatypes.PrivateKey,
+                                 chain_id: int=None) -> Tuple[int, int, int]:
```
Comment (pipermerriam) [inline eth/utils/transactions.py]: candidate for eth-typing `VRS = NewType("VRS", Tuple[int, int, int])`

also, `UINT256 = NewType("UINT256", int)` for these values that are specifically bounded within the 256 bit integer space.

Comment (cburgdorf) [reply eth/utils/transactions.py]: So `VRS` would then be a `VRS = NewType("VRS", Tuple[UINT256, UINT256, UINT256])`?

Comment (pipermerriam) [reply eth/utils/transactions.py]: Yes, though `v` *might* have a lower ceiling than the others (not sure what it is, might be a single byte).

Comment (cburgdorf) [reply eth/utils/transactions.py]: Don't talk low level math with me! What does `v` even stand for? Vitalik? ;)

File: eth/typing.py (hunk 2 above)
Comment (cburgdorf) [reply eth/typing.py]: Yep, they let us avoid sprinkling the code with `assert` statements https://github.com/ethereum/py-evm/pull/1404#discussion_r226073247

File: eth/utils/transactions.py (hunk 5 above)
Comment (pipermerriam) [reply eth/utils/transactions.py]: https://bitcoin.stackexchange.com/questions/38351/ecdsa-v-r-s-what-is-v/38909

`v` is the recovery id and looks to be reasonably bounded within a single byte (2**8 or `UINT8`)

**Comment by cburgdorf** [comment]
@Bhargavasomu thanks again for doing the bulk of the work! You may want to check out the few extra commits on top just to get a feel what kind of changes were made.

File: tox.ini [hunk 6]
```diff
@@ -45,7 +45,7 @@ commands=
     native-state-metropolis: pytest {posargs:tests/json-fixtures/test_state.py -k Metropolis}
     lightchain_integration: pytest --integration {posargs:tests/trinity/integration/test_lightchain_integration.py}
 
-deps = .[p2p,trinity,eth-extra,test]
+deps = .[eth,p2p,trinity,eth-extra,test]
```
Comment (carver) [inline tox.ini]: This is confusing to me: `.` should already be installing those packages:
https://github.com/cburgdorf/py-evm/blob/db8bf2e1003bb18542a3e36ff46fbffabfddc808/setup.py#L102

File: setup.py [hunk 7]
```diff
@@ -12,6 +12,7 @@
         "eth-typing>=2.0.0,<3.0.0",
         "eth-utils>=1.3.0b0,<2.0.0",
         "lru-dict>=1.1.6",
+        "mypy_extensions>=0.4.1,<1.0.0",
```
Comment (carver) [inline setup.py]: Feels a little weird, like including `flake8` in the distribution. Don't feel strongly about it though.

Comment (cburgdorf) [reply setup.py]: At the end of the day, `mypy_extensions` is a library as any other and we are using it now for things like `TypedDict`. Notice that this is not just a type hint per se. We have to have it in order for code such as the following to work.

https://github.com/ethereum/py-evm/blob/7ad4259f574db2214b03da3e8781dca5bbac04e5/eth/typing.py#L20-L25

File: tox.ini (hunk 6 above)
Comment (cburgdorf) [reply tox.ini]: Thanks for pointing that out, now I'm confused as well. Let's see https://github.com/ethereum/py-evm/pull/1417

**Comment by Bhargavasomu** [comment]
@cburgdorf I noticed that you have moved some type definitions to `eth.typing` and importing from there, which is fancy BTW. I haven't known we could do this, definitely will use this method for frequently invoking type definitions

**Comment by cburgdorf** [comment]
Yep, and in the long run we might move some if not all to https://github.com/ethereum/eth-typing

**Pull request merged**